### ISSUE TITLE
Self-Fizzling Turrets

### ISSUE THREAD

**Issue by Endermage77** [BEEmod/BEE2-items#1612]
There are some situations in chambers where large numbers of deactivated turrets could be used by cunning players to create unintended shortcuts; for example, breaking a turret-denial area by hiding behind a wall of deactivated turrets, or possibly constructing piles out of them to climb up to previously unreachable areas.

The self-fizzling turret would fizzle itself upon being deactivated, to prevent this sort of silliness happening.

**Comment by Byzarru** [comment]
Map designers should foresee this happening, tho.

**Comment by Endermage77** [comment]
Also...
It would help keep things less laggy?

**Comment by Byzarru** [comment]
Is that usually such a problem, tho?

**Comment by libbybapa** [comment]
Byzarru, are leftover turret carcasses a problem, ABSOLUTELY.  Often times there is not any elegant way to prevent shortcuts that they can cause from being used for climbing or for breaking laser beams.  The comment that a map designer should foresee this happening is a bit silly.  Of course a map designer can foresee this, this the request for being able to fizzle them...  To say that a map designer should redesign to avoid those issues is the same as saying a map designer should redesign to avoid all the issues that the various BEE objects can prevent and that none of the BEE objects are actually necessary...

Endermage77, there are ways to already do what you ask.  I have two different maps where I have hidden fizzlers in the wall (wall built back up with partial blocks) and then when the player gets to the intended point of bypassing the turrets, a trigger activates the fizzler for a single pulse and fizzles them all.  I used that method in this map:  http://steamcommunity.com/sharedfiles/filedetails/?id=762813279

and this one:  http://steamcommunity.com/sharedfiles/filedetails/?id=808100390

If you want additional details on how to set it up, let me know.

**Comment by TeamSpen210** [comment]
Should they also fizzle after being coated in Repulsion Gel?

**Comment by libbybapa** [comment]
I would envision this as a property that you could select (e.g. start enabled).  If enabled, then once the turret was deactivated it would activate its 'self-destruct' mechanism.  Similarly, make the turret connectable so that when whatever is connected to it (e.g. button, trigger, etc...) then the self-destruct would be activated as well.  This ability would be a very easy and elegant way to deal with the issues that result from undesired uses of turrets.  

...so, if the turret was deactivated by being hit with repulsion gel and start enabled was checked, then yes, it would self-destruct.

**Comment by libbybapa** [comment]
Instead of fizzling, another really nice alternative would be to have them explode like when they are hit with a laser.

**Comment by TeamSpen210** [comment]
I'd want to reserve the turret item input for a different purpose - you can tell them to shut down, so they ignore players. That might be an interesting gameplay element.

In regards to Repulsion Gel, there's 3 outputs for turret defeat - `OnTipped` (activated when the turret finishes randomly firing and closes permanently), `OnBouncePainted` (fired as soon as it's painted with blue gel), and `OnExploded` (after it explodes from lasers). The gel output would fire immediately, but the tipping output fires only after it actually shoots off all its bullets.

**Comment by libbybapa** [comment]
Being unable to trigger the self-destruct without deactivating the turret would still allow players to move them around (still alive) and set them places to use for climbing or use them to break laser beams.

**Comment by libbybapa** [comment]
How about Input A and Input B?

**Comment by Luke18033** [comment]
You can't use turrets to break lasers, they'll just explode.

**Comment by libbybapa** [comment]
WTF?  Yes, they explode but they break the laser beam first.  How about trying something before spouting nonsense.

**Comment by Luke18033** [comment]
I know, I'm just saying you can't block them for long...

**Comment by libbybapa** [comment]
And yet, breaking a laser beam for just a moment can introduce a shortcut.  It did in a map of mine...

**Comment by libbybapa** [comment]
Do you need further explanation/examples or can you accept that as fact without me needing to work that hard?

**Comment by Luke18033** [comment]
I understand that turrets can cause unintended solutions, and I'm not against adding this. I'm just pointing out that turrets can only block lasers for about a second, which wouldn't cause an issue in *most* maps.

**Comment by libbybapa** [comment]
You're right.  Most maps don't even use turrets.  The combination of a turret and a laser beam that would cause a shortcut if broken would be even more obscure.  And yet, it does happen.  Climbing is a similar issue.  The geometry has to be just right to give that portal shot or extra height to get where you shouldn't, but again, it happens.  Both have happened to me and I have seen maps by other players where one or the other of those can be used. 

I can think of one offhand The Delay by themoddoc where he initially had a turret.  He also had this setup:  
![20170220182716_1](https://cloud.githubusercontent.com/assets/19520567/23147200/50410ef4-f79a-11e6-8df9-7ab36cc23660.jpg)

The catcher spawns the reflective cube you see on the island.  The intended solution was to use a funnel to move a cube to block the laser and respawn the cube.  A large chunk of the puzzle (acquiring the cube) could be bypassed by placing the turret in the funnel instead.  It broke the beam only for a moment, but that moment was all that was needed to respawn the cube.  He has since removed the turret.

On the map where I had the issue, the catcher activated a fizzler.  Again, the intended was to funnel a cube to break the laser and allow a portal shot into another room.  It was possible to drop a turret into the funnel, though, and break the laser just long enough to get the shot off.

**Comment by TeamSpen210** [comment]
A solution to that problem is to use the Deactivation Delayer. That imposes a delay on the output, so the laser must remain blocked for a certain amount of time. It would ignore a brief blockage from turrets or the like.

**Comment by libbybapa** [comment]
I tried the activation delayer on my map, but it did not solve the issue.  It's a bit hard to explain.  I had 6 turrets.  Any/all of them could be dropped into the funnel.  If one slipped past the laser without burning up, it was skinny enough that the beam would not hit it.  A second turret would rest against the first and explode breaking the beam for a moment.  For whatever reason the proximity of the unexploded turret would show the beam going past, but the catcher would not activate.  I can make a mock-up map to demonstrate the issue if you would like but a delayer will only fix it 50% of the time.

**Comment by Luke18033** [comment]
Oh, so that turret-laser bug isn't just a Hammer issue...

**Comment by libbybapa** [comment]
I don't know Hammer.  I spent quite a bit of time pondering my issue and working with the deactivation delayer and was quite frustrated when that did not actually work.  As I mentioned above, I finally hid a fizzler in the wall (actually I used a matter inquisition field so it would not fizzle the player's portals) and used a combination of trigger, two hidden lasers, and a hidden relay to fizzle the turrets when the player got to that part of the map.  That map is here:  http://steamcommunity.com/sharedfiles/filedetails/?id=762813279

**Comment by TeamSpen210** [comment]
Is it a bug with the item, or an issue with the laser not activating again?

**Comment by libbybapa** [comment]
I just made up a test chamber and a vid.  https://youtu.be/TDDE8-UptXA  The catcher acts like the laser is not hitting it if the turret is near the laser.  The turret will not burn up if its leg is all that is hit by the laser for a moment as it goes by.  It will then turn off the catcher permanently even with a delayer added.  Notice in the vid that the turret is still alive and well.  It doesn't even take two turrets (I wasn't remembering correctly).

**Comment by libbybapa** [comment]
In my Big Push map linked above, if you noclip to where the turrets are under the exit, there are triggers added at the portalable surfaces so that when the player comes through a portal into that area, the turrets all fizzle.  It's actually quite funny.

**Comment by TeamSpen210** [comment]
That is bizarre. This won't help with it either, the turret is still active and alive.

**Comment by Luke18033** [comment]
It's a Valve bug, it happens in Hammer maps too.

**Comment by libbybapa** [comment]
That's why I was asking for the ability to connect an activating object or more specifically trigger to the turret to cause it to self-destruct.  With the ability to connect a trigger, you can place the trigger or triggers at specific points (on top of the turrets) and cause the turrets to self-destruct before they can be used in that unintended manner.

**Comment by libbybapa** [comment]
Here's a vid showing the way I have dealt with it in the two maps I linked earlier:  https://youtu.be/07rv0jeJM7c

It's heavy on the objects (2 lasers, 2 relays, 1 fizzler, 1 trigger) and a pain to create, but it works seamlessly.

**Comment by Byzarru** [comment]
TANGENT TIME!
Now I present _my_ real concern. I'm a big supporter of the "[duck test](https://en.wikipedia.org/wiki/Duck_test)." Turrets don't self-fizzle in the campaign or a majority of maps in the workshop. Thus, identical turrets in a BEE extension shouldn't be able to self-fizzle. This will likely confuse players for the first time, and be a pleasant surprise to some. The problem is that mapmakers and players alike are not always helpful or savvy. Many players do not actually read the workshop descriptions, either out of apathy, loading the map from Portal 2's native feed, or both; likewise, many mappers don't give helpful descriptions of items or concepts in their maps. With the variety of possibilities and concepts on the Workshop, items should be clear and obvious to their purpose after being tested the first time around. Turrets shouldn't be able to self-fizzle if they appear as they do in the campaign, or the rest of the Workshop, because players don't expect a self-fizzling Turret. That is not how they behave in those places. Give it another color, or an emblem, and this becomes a new turret class. It's obvious now that these turrets destroy themselves after an initial test, without needing a Workshop description that either won't be written or read. I think that this would be possible to implement without making it an entirely new item (it could be the Start Open option, leaving Start Enabled open for turret activation).

**Comment by TeamSpen210** [comment]
I agree, which is the other reason I'm reluctant to implement that. It's fine for the Turret Dropper turrets, since you can see them immediately being replaced (and probably respawning).

**Comment by libbybapa** [comment]
I don't see any issue with it at all.  No explanation is needed.  You place the trigger at the location where the player would need to get to in order to destroy the turrets.  Once the fizzle, the cause/effect are obvious and also irrelevant.

**Comment by TeamSpen210** [comment]
Not really, especially if it's from a trigger. People might consider it as a bug even.

**Comment by libbybapa** [comment]
I just don't see that as an issue.  The shortcuts introduced from unintended uses of turrets certainly are an issue, though.  IF the spontaneous combustion of turrets is seen as an issue, then add a turret voice line that is something to the effect of "Self-destruct activated" and then have it fizzle.  I can't imagine that there is any player with the mental acuity to actually load the game who wouldn't figure out that it was intended behavior when the turret fizzled.

**Comment by Byzarru** [comment]
It's quite possible that you don't fall in the typical Workshop userbase, which accounts for a good number of users of BEE as well.

**Comment by Endermage77** [comment]
Actually, there is a problem with confusion. 
I have a couple of ideas for distinguishment:
Assuming you would like self-destruct by activation, you could 
1. Print a signage symbol on the turret if the connection visibility is set to "Signage,"
2. Have an actual wire hooking into it for "Wire" connection,
1 & 2 go together, by the way.
3. Retexture it as said above,
4. Trigger a special voiceline as said above,

Or, if none of those make it, at the very least make the mapmaker do a sort of mini-test for prepping players to test with self-destructing turrets.

**Comment by Byzarru** [comment]
1. That's a good idea. What if they're placed on glass, though? Antlines and signages aren't coded to work on glass.
2. I'm a little confused there. Would this be like a tripwire (visually, of course)?
3. I'm thinking of this as a turret with a danger stripe ring around the 'eye', and a fizzle icon on its sides. If my idea regarding another turret issue comes along (such as specially skinned stealth and superpowered military grade turrets) or TNT turrets are implemented, they should have the same options available to the standard turrets. Visual cues need to be universal.
4. Which would require the antechamber with the turret.
5. Sturgeon's law affects us all. You can't make mapmakers do anything; just give tips and advice and hope for the best.

**Comment by libbybapa** [comment]
"It's quite possible that you don't fall in the typical Workshop user base, which accounts for a good number of users of BEE as well."

I'm quite sure that's true.  I'm also sure that anyone successfully playing (the majority of) my maps is in the top few percentile for intelligence.  I doubt a fizzling turret would confuse or daunt anyone who would not be totally stymied by the logic of test itself.  I don't tend to pander to the lowest common denominator, but I see we differ in that.

"Sturgeon's law affects us all. You can't make mapmakers do anything; just give tips and advice and hope for the best."

I think that leaving the usage to the map-maker's discretion is a good thing.  Anyway... it would be very helpful functionality if added regardless of whether other skins, connection visibility, or other visual/auditory cues are added.  Without it, though, I'm sure I'll get by just fine.

**Comment by Endermage77** [comment]
You'd be surprised.

**Comment by Endermage77** [comment]
Looking back at this conversation, I see there was some confusion as to the interpretation of the post.

The conversation above is talking about turrets that are fizzled by signals, when in actuality the post was originally about turrets that fizzled themselves _automatically_ when you killed them.
